File: typescript/exercises.ts
import { error } from "node:console";
import { open } from "node:fs/promises";

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative");
  }
  let counts: Map<bigint, bigint> = new Map();
  let remaining = amount;
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination);
    remaining %= denomination;
  }
  return counts;
}

export function firstThenApply<T, U>(
  sequence: T[],
  predicate: (element: T) => boolean,
  apply: (item: T) => U
): U | undefined {
  for (const item of sequence) {
    if (predicate(item)) {
      return apply(item);
    }
  }
  return undefined;
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  for (let power = 1n; ; power *= base) {
    yield power;
  }
}

export async function meaningfulLineCount(filename: string): Promise<number> {
  let count = 0;
  const file = await open(filename, "r");
  for await (const line of file.readLines()) {
    // Note that readLines will autoclose the file, yay
    const trimmed = line.trim();
    if (trimmed && !trimmed.startsWith("#")) {
      count++;
    }
  }
  return count;
}

interface Sphere {
  readonly kind: "Sphere";
  readonly radius: number;
}

interface Box {
  readonly kind: "Box";
  readonly width: number;
  readonly length: number;
  readonly depth: number;
}

export type Shape = Sphere | Box;

export function surfaceArea(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return (
        2 * shape.width * shape.depth +
        2 * shape.length * shape.width +
        2 * shape.depth * shape.length
      );
    case "Sphere":
      return 4 * Math.PI * shape.radius ** 2;
  }
}

export function volume(shape: Shape): number {
  switch (shape.kind) {
    case "Box":
      return shape.width * shape.length * shape.depth;
    case "Sphere":
      return (4 / 3) * Math.PI * shape.radius ** 3;
  }
}

export interface BinarySearchTree<T> {
  size(): number;
  insert(value: T): BinarySearchTree<T>;
  contains(value: T): boolean;
  inorder(): Iterable<T>;
}

export class Empty<T> implements BinarySearchTree<T> {
  size(): number {
    return 0;
  }

  insert(value: T): BinarySearchTree<T> {
    return new Node<T>(value, new Empty(), new Empty());
  }

  contains(value: T): boolean {
    return false;
  }

  *inorder(): Iterable<T> {}

  toString(): string {
    return "()";
  }

  constructor() {
    Object.freeze(this);
  }
}

class Node<T> implements BinarySearchTree<T> {
  left: BinarySearchTree<T>;
  right: BinarySearchTree<T>;
  value: T;

  constructor(value: T, left: BinarySearchTree<T>, right: BinarySearchTree<T>) {
    this.left = left;
    this.right = right;
    this.value = value;
    Object.freeze(this);
  }

  insert(value: T): BinarySearchTree<T> {
    if (value === this.value) {
      return this;
    } else if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right);
    } else {
      return new Node(this.value, this.left, this.right.insert(value));
    }
  }

  contains(value: T): boolean {
    if (value == this.value) {
      return true;
    } else if (value < this.value) {
      return this.left.contains(value);
    } else {
      return this.right.contains(value);
    }
  }

  size(): number {
    return 1 + this.left.size() + this.right.size();
  }

  *inorder(): Iterable<T> {
    yield* this.left.inorder();
    yield this.value;
    yield* this.right.inorder();
  }

  toString(): string {
    return `(${this.left.toString().replace("()", "")}${this.value}${this.right.toString().replace("()", "")})`;
  }
}

File: ocaml/exercises.ml
(* to compile:
  (& opam env) -split '\r?\n' | ForEach-Object { Invoke-Expression $_ }
  ocamlc .\exercises.ml .\exercises_test.ml; .\camlprog.exe 
  *)

exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let first_then_apply array predicate consumer = 
  match List.find_opt predicate array with
  | None -> None
  | Some x -> consumer x;;

let powers_generator base = 
  let rec generate_from power () =
    Seq.Cons(power, generate_from (power * base))
  in
  generate_from 1;;

let meaningful_line_count fileName = 
  let meaningful_line line = 
    let trimmed = String.trim line in
    String.length trimmed > 0 && not (String.get trimmed 0 = '#')
  in
  let the_file = open_in fileName in
  let finally () = close_in the_file in
  let rec count_lines count = 
    try
      let line = input_line the_file in
      if meaningful_line line then
        count_lines (count + 1)
      else
        count_lines count
    with
      End_of_file -> count
  in
  Fun.protect ~finally (fun () -> count_lines 0);;

type shape = 
  | Sphere of float 
  | Box of float * float * float

let volume s = 
  match s with
  | Sphere r -> Float.pi *. (r ** 3.) *. 4. /. 3.
  | Box (l, w, h) -> l *. w *. h;;

let surface_area s =
  match s with
  | Sphere r -> Float.pi *. 4. *. (r ** 2.)
  | Box (l, w, h) -> 2. *. l *. w +. 2. *. l *. h +. 2. *. h *. w

type 'a binary_search_tree = 
  | Empty
  | Node of 'a binary_search_tree * 'a * 'a binary_search_tree

let rec size tree = 
  match tree with
  | Empty -> 0
  | Node (left, _, right) -> 1 + size left + size right;;

let rec insert item tree =
  match tree with
  | Empty -> Node (Empty, item, Empty)
  | Node (left, value, right) ->
    if item < value then
      Node (insert item left, value, right)
    else if item > value then
      Node(left, value, insert item right)
    else
      Node(left, item, right)

let rec contains value tree =
  match tree with 
  | Empty -> false
  | Node (left, v, right) ->
    if value = v then
      true
    else if value < v then
      contains value left
    else
      contains value right;;

let rec inorder tree = 
  match tree with
  | Empty -> []
  | Node (left, v, right) -> inorder left @ [v] @ inorder right;;